README.md
Maildir Deduplicate
===================

Command-line tool to deduplicate mails from a set of maildir folders.


Usage
-----

```
Usage: __init__.py [OPTIONS] [MAILDIR [MAILDIR ...]]

Detect/remove duplicates from maildir folders

Options:
  -h, --help            show this help message and exit
  -d, --remove-smaller  Remove all but largest duplicate in each duplicate set
  -r REGEXP, --remove-matching=REGEXP
                        Remove duplicates whose file path matches REGEXP
  -R REGEXP, --remove-not-matching=REGEXP
                        Remove duplicates whose file path does not match
                        REGEXP
  -o, --remove-older    Remove all but the newest duplicate (determined by
                        ctime) in each duplicate set
  -O, --remove-newer    Remove all but the oldest duplicate (determined by
                        ctime) in each duplicate set
  -n, --dry-run         Don't actually remove anything; just show what would
                        be removed.
  -s, --show-diffs      Show diffs between duplicates even if they're within
                        the thresholds
  -i, --message-id      Use Message-ID header as hash key (not recommended -
                        the default is to compute a digest of the whole header
                        with selected headers removed)
  -S BYTES, --size-threshold=BYTES
                        Specify maximum allowed difference between size of
                        duplicates. Default is 512; set -1 for no threshold.
  -D BYTES, --diff-threshold=BYTES
                        Specify maximum allowed size of unified diff between
                        duplicates. Default is 768; set -1 for no threshold.
  -H, --hash-pipe       Take a single mail message texted piped from STDIN and
                        show its canonicalised form and hash thereof. This is
                        useful for debugging why two messages don't have the
                        same hash when you expect them to (or vice-versa).
```


Details
-------

This script reads all mails in a given list of maildir folders and
subfolders, then automatically detects, lists, and optionally
deletes any duplicate mails.

Duplicate detection is done by cherry-picking certain headers, in
some cases doing some minor tweaking of the values to reduce them
to a canonical form, and then computing a digest of those headers
concatenated together.

Note that we deliberately limit this to certain headers due to the
effects that mailing list software can have on not only the mail
header but the body; it can potentially:

  * append a footer to a list body, thus changing the `Content-Length` header

  * create a new path described by the `Received` headers which would not be contained in any copy of the mail saved locally at the time it was sent to the list

  * munge the `Reply-To` header even though it's a bad idea

  * add plenty of other random headers which a copy saved locally at sending-time would not have, such as `X-Mailman-Version`, `Precedence`, `X-BeenThere`, `List-*`, `Sender`, `Errors-To`, and so on.

  * add a prefix to the `Subject` header

Another difficulty is the lack of guarantee that `Message-ID` is
unique or even present.  Yes, certain broken mail servers which
must remain nameless are guilty of this :-(

For added protection against accidentally removing mails due to
false positives, duplicates are verified by comparing body sizes
and also diff'ing the contents.  If the sizes or contents differ
by more than a threshold, they are not counted as duplicates.

So far, it was tested on:

  * MacOS X 10.6 with Python 2.6.2,
  * Linux with Python 2.6.0 and 2.7.2.


Authors & Contributors
----------------------

  * [Kevin Deldycke](https://github.com/kdeldycke)
  * [Adam Spiers](https://github.com/aspiers)
  * [Ben Reser](https://github.com/breser)
  * [Marcel Martin](https://github.com/marcelm)
  * [reedog117](https://github.com/reedog117)
  * [Tristan Henderson](https://github.com/tnhh)
  * [Matija Nalis](https://github.com/mnalis)


History
-------

This script was [initially released in 2010](http://kevin.deldycke.com/2010/08/maildir-deduplication-script-python/), and was living in a [messy GitHub repository](https://github.com/kdeldycke/scripts). After some years, the script basically outgrew its initial intent, and [moved in 2013 to its own repository](http://kevin.deldycke.com/2013/06/maildir-deduplicate-moved/).


License
-------

This software is licensed under the [GNU General Public License v2 or later (GPLv2+)](https://github.com/kdeldycke/maildir-deduplicate/blob/master/LICENSE).


CHANGES.rst
ChangeLog
=========

`3.0.1 (2021-01-25) <https://github.com/kdeldycke/mail-deduplicate/compare/v3.0.0...v3.0.1>`_
---------------------------------------------------------------------------------------------

* Add explicit warning in CLI output to warn about 3.x branch deprecation.
  Refs #180.


`3.0.0 (2020-09-03) <https://github.com/kdeldycke/mail-deduplicate/compare/v2.2.0...v3.0.0>`_
---------------------------------------------------------------------------------------------

.. DANGER::
   As of ``v3.0.0``, the project has been renamed to *Mail Deduplicate* and is
   now available on:

   * Github at https://github.com/kdeldycke/mail-deduplicate
   * PyPi at https://pypi.org/project/mail-deduplicate

* Rename project from ``maildir-deduplicate`` to ``mail-deduplicate``.
* Rename ``master`` branch to ``main``.
* Add support for mboxes. Closes #48.
* Remove requirement on ``-s``/``--strategy`` parameter, to let mails being
  grouped into duplicate sets without any removal action, effectively acting as
  a second-level dry-run.


`2.2.0 (2020-09-03) <https://github.com/kdeldycke/mail-deduplicate/compare/v2.1.0...v2.2.0>`_
------------------------------------------------------------------------------------------------

.. warning::
   This is the last version of the project known under the name *Maildir
   Deduplicate* (a.k.a. ``maildir-deduplicate``).

* Add disclaimer to prepare project name change.
* Fix Header being object instead of string (#61).
* Make body_lines conversion more reliable.
* Fix bugs in counter statistics (#45, #57).
* Add Message-ID as a header to check.
* Fix ``UnicodeDecodeError``. Closes #53 and #55.
* Bump requirement to ``click_log >= 0.2.0``. Closes #58, #59 and #60.
* Replace ``nose`` by ``pytest``.
* Only notify by mail of test failures.
* Drop support of Python 3.3.


`2.1.0 (2017-01-13) <https://github.com/kdeldycke/mail-deduplicate/compare/v2.0.1...v2.1.0>`_
------------------------------------------------------------------------------------------------

* Fix rendering of changelog link in RST.
* Show selected log level in debug mode.
* Test builds against Python 3.6 and 3.7-dev, and most recent PyPy targetting
  Python 2.7 and 3.3.
* Bump requirement to ``readme_renderer >= 16.0``.
* Skip sets with unparseable mails because of incorrect encoding. Closes #47,
  #41 and #39.
* Require the ``-s``/``--strategy`` CLI parameter to be set. Closes #44.


`2.0.1 (2016-11-28) <https://github.com/kdeldycke/mail-deduplicate/compare/v2.0.0...v2.0.1>`_
------------------------------------------------------------------------------------------------

* Reject mails without ``Message-ID`` headers if ``--message-id`` option is
  active.
* Add a class to holds global config.
* Print more debug info in unittests when CLI produce tracebacks.
* Always check for package metadata in Travis CI jobs.
* Fix package's long description against PyPi rendering constraints.
* Add link to full changelog in package's long description.


`2.0.0 (2016-11-13) <https://github.com/kdeldycke/mail-deduplicate/compare/v1.3.0...v2.0.0>`_
------------------------------------------------------------------------------------------------

* Refactor and redefine all removal strategies. Closes #40.
* Add a new ``-t``/``--time-source`` CLI parameter to point to the canonical
  source of a mail's timestamp.
* Rename ``-s``/``--show-diffs`` parameter to ``-d``/``--show-diff``.
* Rename ``-D``/``--diff-threshold`` parameter to
  ``-C``/``--content-threshold``.
* Add new ``-s`` shorthand to ``--strategy`` parameter.
* Re-implement diff threshold options. Closes #43.
* Re-implement statistics. Closes #42.
* Normalize and canonicalize all paths to deduplicate on the fly mails pointing
  to the same file.
* Build documentation via Sphinx.
* Add ``test`` and ``develop`` dependencies.
* Move coverage config to ``setup.cfg``.
* Replace ``pep8`` package by ``pycodestyle``.
* Enforce ``pycodestyle`` checks in Travis CI jobs.
* Remove popularity badge: PyPI download counters are broken and no longer
  displayed.
* Test production of packages in Travis CI jobs.
* Move development and packaging documentation to Sphinx. Closes #22.
* Make wheels generated under Python 2 environnment available for Python 3 too.
* Let unittests generates their own fake and temporary maildirs.
* Print CLI output in unittests.
* Use generic factory to produce mail fixtures in unittests.
* Only show latest changes in the long description of the package instead of
  the full changelog.


`1.3.0 (2016-08-11) <https://github.com/kdeldycke/mail-deduplicate/compare/v1.2.0...v1.3.0>`_
------------------------------------------------------------------------------------------------

* User-friendly progress bar.
* Decrease memory usage on large datasets. Closes #19, #8 and #3.
* Attempt several encodings when parsing message body. Closes #32.
* Fixed comparison issue in Python 3. Closes #34.
* Add a set of basic deduplication unittests.


`1.2.0 (2016-03-29) <https://github.com/kdeldycke/mail-deduplicate/compare/v1.1.0...v1.2.0>`_
------------------------------------------------------------------------------------------------

* Use logger to output messages to the user.
* Activate tests on Python 3.3, PyPy and PyPy3.
* Fix date parsing. See #33.
* Fix decoding of unicode header value. Closes #24 and #32.


`1.1.0 (2016-01-10) <https://github.com/kdeldycke/mail-deduplicate/compare/v1.0.2...v1.1.0>`_
------------------------------------------------------------------------------------------------

* Add Python 3.4 and 3.5 support. Closes #30.
* Add default ``isort`` configuration.
* Remove hackish default encoding forcing in main code.


`1.0.2 (2015-12-22) <https://github.com/kdeldycke/mail-deduplicate/compare/v1.0.1...v1.0.2>`_
------------------------------------------------------------------------------------------------

* Fix ``AttributeError`` on message instance. Closes #28.

mail_deduplicate/cli.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2010-2017 Kevin Deldycke <kevin@deldycke.com>
#                         and contributors.
# All Rights Reserved.
#
# This program is Free Software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

from __future__ import absolute_import, division, print_function

import logging
import os
import re
import time

import click
import click_log

from . import (
    DEFAULT_CONTENT_THRESHOLD,
    DEFAULT_SIZE_THRESHOLD,
    DELETE_MATCHING_PATH,
    DELETE_NEWER,
    DELETE_NEWEST,
    DELETE_NON_MATCHING_PATH,
    DELETE_OLDER,
    DELETE_OLDEST,
    MD_SUBDIRS,
    STRATEGIES,
    TIME_SOURCES,
    Config,
    __version__,
    logger
)
from .deduplicate import Deduplicate
from .mail import Mail

click_log.basic_config(logger)

logger.warning("XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX")
logger.warning("XXX                                                               XXX")
logger.warning("XXX         THIS VERSION OF mdedup IS COMPLETELY UNSUPPORTED      XXX")
logger.warning("XXX PLEASE UPDATE TO PYTHON3.6 OR LATER AND mdedup 6.1.x OR LATER XXX")
logger.warning("XXX                                                               XXX")
logger.warning("XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX")
logger.warning("")
time.sleep(5)

@click.group(invoke_without_command=True)
@click_log.simple_verbosity_option(
    logger, default='INFO', metavar='LEVEL',
    help='Either CRITICAL, ERROR, WARNING, INFO or DEBUG. Defaults to INFO.')
@click.version_option(__version__)
@click.pass_context
def cli(ctx):
    """ CLI for mbox and maildir content analysis and deletion. """
    level = logger.level
    try:
        level_to_name = logging._levelToName
    # Fallback to pre-Python 3.4 internals.
    except AttributeError:
        level_to_name = logging._levelNames
    level_name = level_to_name.get(level, level)
    logger.debug('Verbosity set to {}.'.format(level_name))

    # Print help screen and exit if no sub-commands provided.
    if ctx.invoked_subcommand is None:
        click.echo(ctx.get_help())
        ctx.exit()

    # Load up global options to the context.
    ctx.obj = {}


def validate_regexp(ctx, param, value):
    """ Validate and compile regular expression. """
    if value:
        try:
            value = re.compile(value)
        except ValueError:
            raise click.BadParameter('invalid regular expression.')
    return value


def validate_mail_sources(ctx, param, value):
    """ Check that files are mboxes and folders are maildirs. """
    for path in value:
        # Validates folder is a maildir.
        if os.path.isdir(path):
            for subdir in MD_SUBDIRS:
                if not os.path.isdir(os.path.join(path, subdir)):
                    raise click.BadParameter(
                        '{} is not a maildir folder (missing {!r} '
                        'sub-directory).'.format(path, subdir))
        # Validates file is an mbox.
        else:
            if not os.path.isfile(path):
                raise click.BadParameter(
                    '{} is not an mbox file.'.format(path))
            # TODO: Insert here early checks on mbox file format.
    return value


@cli.command(short_help='Deduplicate mboxes and/or maildirs content.')
@click.option(
    '-s', '--strategy', type=click.Choice(STRATEGIES),
    help='Deletion strategy to apply within a subset of duplicates. If not '
    'set duplicates will be grouped and counted but no removal will happens.')
@click.option(
    '-t', '--time-source', type=click.Choice(TIME_SOURCES),
    help="Source of a mail's reference time. Required in time-sensitive "
    'strategies.')
@click.option(
    '-r', '--regexp', callback=validate_regexp, metavar='REGEXP',
    help='Regular expression against a mail file path. Required in '
    'delete-matching-path and delete-non-matching-path strategies.')
@click.option(
    '-n', '--dry-run', is_flag=True, default=False,
    help='Do not actually delete anything; just show what would be removed.')
@click.option(
    '-i', '--message-id', is_flag=True, default=False,
    help='Only use the Message-ID header as a hash key. Not recommended. '
    'Replace the default behavior consisting in deriving the hash from '
    'several headers.')
@click.option(
    '-S', '--size-threshold', type=int, metavar='BYTES',
    default=DEFAULT_SIZE_THRESHOLD,
    help='Maximum allowed difference in size between mails. Whole subset of '
    'duplicates will be rejected above threshold. Set to -1 to not allow any '
    'difference. Defaults to {} bytes.'.format(DEFAULT_SIZE_THRESHOLD))
@click.option(
    '-C', '--content-threshold', type=int, metavar='BYTES',
    default=DEFAULT_CONTENT_THRESHOLD,
    help='Maximum allowed difference in content between mails. Whole subset '
    'of duplicates will be rejected above threshold. Set to -1 to not allow '
    'any difference. Defaults to {} bytes.'.format(DEFAULT_CONTENT_THRESHOLD))
@click.option(
    '-d', '--show-diff', is_flag=True, default=False,
    help='Show the unified diff of duplicates not within thresholds.')
# TODO: add a show-progress option.
@click.argument(
    'mail_sources', nargs=-1, callback=validate_mail_sources,
    metavar='MBOXES/MAILDIRS', type=click.Path(exists=True, resolve_path=True))
@click.pass_context
def deduplicate(
        ctx, strategy, time_source, regexp, dry_run, message_id,
        size_threshold, content_threshold, show_diff, mail_sources):
    """ Deduplicate mails from a set of either mbox files or maildir folders.

    Run a first pass computing the canonical hash of each encountered mail from
    their headers, then a second pass to apply the deletion strategy on each
    subset of duplicate mails.

    \b
    Removal strategies for each subsets of duplicate mails:
      - delete-older:    Deletes the olders,    keeps the newests.
      - delete-oldest:   Deletes the oldests,   keeps the newers.
      - delete-newer:    Deletes the newers,    keeps the oldests.
      - delete-newest:   Deletes the newests,   keeps the olders.
      - delete-smaller:  Deletes the smallers,  keeps the biggests.
      - delete-smallest: Deletes the smallests, keeps the biggers.
      - delete-bigger:   Deletes the biggers,   keeps the smallests.
      - delete-biggest:  Deletes the biggests,  keeps the smallers.
      - delete-matching-path: Deletes all duplicates whose file path match the
      regular expression provided via the --regexp parameter.
      - delete-non-matching-path: Deletes all duplicates whose file path
      doesn't match the regular expression provided via the --regexp parameter.

    Deletion strategy on a duplicate set only applies if no major differences
    between mails are uncovered during a fine-grained check differences during
    the second pass. Limits can be set via the threshold options.
    """
    # Print help screen and exit if no mail source provided.
    if not mail_sources:
        click.echo(ctx.get_help())
        ctx.exit()

    # Validate exclusive options requirement depending on strategy.
    requirements = [
        (time_source, '-t/--time-source', [
            DELETE_OLDER, DELETE_OLDEST, DELETE_NEWER, DELETE_NEWEST]),
        (regexp, '-r/--regexp', [
            DELETE_MATCHING_PATH, DELETE_NON_MATCHING_PATH])]
    for param_value, param_name, required_strategies in requirements:
        if strategy in required_strategies:
            if not param_value:
                raise click.BadParameter(
                    '{} strategy requires the {} parameter.'.format(
                        strategy, param_name))
        elif param_value:
            raise click.BadParameter(
                '{} parameter not allowed in {} strategy.'.format(
                    param_name, strategy))

    conf = Config(
        strategy=strategy,
        time_source=time_source,
        regexp=regexp,
        dry_run=dry_run,
        show_diff=show_diff,
        message_id=message_id,
        size_threshold=size_threshold,
        content_threshold=content_threshold,
        # progress=progress,
    )

    dedup = Deduplicate(conf)

    logger.info('=== Phase #1: load mails and compute hashes.')
    for source in mail_sources:
        dedup.add_source(source)

    logger.info('=== Phase #2: deduplicate mails.')
    dedup.run()

    dedup.report()


@cli.command(short_help='Hash a single mail.')
# TODO: Deduplicate option definition.
@click.option(
    '-i', '--message-id', is_flag=True, default=False,
    help='Only use the Message-ID header as a hash key. Not recommended. '
    'Replace the default behavior consisting in deriving the hash from '
    'several headers.')
@click.argument('message', type=click.Path(
    exists=True, dir_okay=False, resolve_path=True))
@click.pass_context
def hash(ctx, message_id, message):
    """ Take a single mail message and show its canonicalised form and hash.

    Mainly used to debug message hashing. Only works with maildirs, not mboxes.
    """
    conf = Config(message_id=message_id)

    # TODO: Update with new mail API.
    mail = Mail(message, conf)

    logger.info(mail.header_text)
    logger.info('-' * 70)
    logger.info('Hash: {}'.format(mail.hash_key))
